cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) 2016 SnappyData, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you
+ * may not use this file except in compliance with the License. You
+ * may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License. See accompanying
+ * LICENSE file.
+ */
+package org.apache.spark.memory
+
+import scala.collection.mutable.ArrayBuffer
+
+import com.gemstone.gemfire.internal.cache.GemFireCacheImpl
+import com.pivotal.gemfirexd.internal.engine.Misc
+import org.apache.spark.SparkConf
+import scala.collection.mutable
+
+import org.apache.spark.storage.{BlockStatus, BlockId}
+
+/**
+  * When there is request for execution or storage memory, critical up and eviction up
+  * events are checked. If they are set, try to free the memory cached by Spark rdds
+  * by calling memoryStore.evictBlocksToFreeSpace. If enough memory cannot be freed,
+  * return the call and let Spark take a corrective action.
+  * In such cases Spark either fails the task or move the current RDDs data to disk.
+  * If the critical and eviction events are not set, it asks the UnifiedMemoryManager
+  * to allocate the space.
+  * @param conf
+  * @param maxMemory
+  * @param numCores
+  */
+class SnappyUnifiedMemoryManager private[memory](
+    conf: SparkConf,
+    override val maxMemory: Long,
+    numCores: Int) extends UnifiedMemoryManager(conf,
+  maxMemory,
+  (maxMemory * conf.getDouble("spark.memory.storageFraction", 0.5)).toLong,
+  numCores) {
+
+  def this(conf: SparkConf, numCores: Int) = {
+    this( conf,
+      SnappyUnifiedMemoryManager.getMaxMemory(conf),
+      numCores)
+  }
+  private val listener = Misc.getMemStoreBooting.thresholdListener()
+
+  private def freeMemory(blockId: Option[BlockId], numBytes: Long,
+      evictedBlocks: mutable.Buffer[(BlockId, BlockStatus)]): Boolean = {
+    val freeMemory = Runtime.getRuntime.freeMemory()
+    if (freeMemory < numBytes) {
+
+      storageMemoryPool.memoryStore.evictBlocksToFreeSpace(blockId,
+        numBytes - freeMemory,
+        evictedBlocks) || !listener.isCritical
```

Comment:
Why does freeMemory() return true when it fails to evict blocks but listener.critical is false?

In SnappyUnifiedMemoryManager. acquireStorageMemory / acquireExecutionMemory, this will mean that when eviction is up, if Spark mem allocation is below spark.memory.fraction, even when no block was evicted memory will be acquired due to calls to super.acquireStorageMemory/super.acquireExecutionMemory.   Do we want Spark to continue acquiring storage memory in such case and only GemXD keeps evicting table data?